text/0000-min-rust-version.md

```diff
@@ -0,0 +1,179 @@
+Rust 1.40 or higher. But if you'll try to build your project with Rust 1.29 cargo
+will issue an error.
+
+`rust` field value will be checked as well. During crate build `cargo` will check
+if all upstream dependencies can be built with the specified MSRV. (i.e. it will
+check if there is exists solution for given crates and Rust versions constraints)
+Yanked crates will be ignored in this process.
+
+Implementing this functionality hopefully will allow us to close the long-standing
+debate regarding whether MSRV bump is a breaking change or not and will allow
+crate authors to feel less restrictive about bumping their crate's MSRV. (though
+it may be a useful convention for post-1.0 crates to bump minor version on MSRV
+bump to allow publishing backports which fix serious issues using patch version)
+
+Note that described MSRV constraints and checks for dependency versions resolution
+can be disabled with the `--no-msrv-check` option.
+
+## Checking MSRV during publishing
+
+`cargo publish` will check that upload is done with a toolchain version specified
+in the `rust` field. If toolchain version is different, `cargo` will refuse to
+upload the crate. It will be a failsafe to prevent uses of incorrect `rust` values
+due to unintended MSRV bumps. This check can be disabled by using the existing
+`--no-verify` option.
+
+## Making `rust` field mandatory
+
+In future (probably in a next edition) we could make `rust` field mandatory for
+a newly uploaded crates. MSRV for older crates will be determined by the `edition`
+field. In other words `edition = "2018"` will imply `rust = "1.31"` and
+`edition = "2015"` will imply `rust = "1.0"`.
+
+`cargo init` will use version of a currently used toolchain.
+
+## `cfg`-based MSRVs
+
+Some crates can have different MSRVs depending on target architecture or enabled
+features. In such cases it can be useful to describe how MSRV depends on them,
+e.g. in the following way:
+```toml
+[package]
+rust = "1.30"
+
+[target.x86_64-pc-windows-gnu]
+rust = "1.35"
+
+[target.'cfg(feature = "foo")']
+rust = "1.33"
+```
+
+All `rust` values in the `target` sections should be equal or bigger to a `rust` value
+specified in the `package` section.
+
+If target condition is true, then `cargo ` will use `rust` value from this section.
+If several target section conditions are true, then maximum value will be used.
+
+## Nightly and stable versions
+
+Some crates may prefer to target only the most recent stable or nightly toolchain.
+In addition to the versions we could allow `stable` and `nightly` values to declare
+that maintainers do not track MSRV for the crate.
+
+For some bleeding-edge crates which experience frequent breaks on Nightly updates
+(e.g. `rocket`) it can be useful to specify exact Nightly version(s) on which
+crate can be built. One way to achieve this is by using the following syntax:
+- auto-select: "nightly" This variant will behave in the same way as "stable", i.e.
```

Comment:
This entire list is poorly formatted for reading, with this line being particularly problematic.